### PR TITLE
Refactor ChannelsView and add filter

### PR DESCRIPTION
**Pull request by broskoski**
channels-filter

### CODE REVIEW

File: apps/user/components/channels/collection.coffee [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+  url: -> "/api/#{@userSlug}/channels?#{params.encode(@params.toJSON())}"
+
+  parse: (data) -> data.channels
+
+  initialize: (models, { @userSlug, @params }) ->
+    throw new Error 'Requires a params model' unless @params?
+    throw new Error 'Requires a userSlug' unless @userSlug?
+
+    @params.on 'change', @prepareAndFetch, @
```
Comment (broskoski) [inline apps/user/components/channels/collection.coffee]: This collection listens to changes on `@params` and fetches.

File: apps/user/components/channels/view.coffee [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    @loadNextPage()  if total - progress < @threshold
+
+  query: ->
+    val = @$('.js-channel-filter').val()
+    if val is '' then null else val 
+
+  searchChannels: =>
+    @state.set loading: true, disabled: false
+
+    @params.set
```
Comment (broskoski) [inline apps/user/components/channels/view.coffee]: Setting `@params` to paginate or search

File: apps/user/components/channels/collection.coffee [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+  prepareAndFetch: ->
+    paginating = contains(keys(@params.changed), 'page') and keys(@params.changed).length is 1
+    options = if paginating then { remove: false } else {}
+    @fetch options
+
+  onlyBlocks: ->
```
Comment (broskoski) [inline apps/user/components/channels/collection.coffee]: `onlyBlocks` lets us skip through the blocks returned in lightbox mode.

File: apps/user/client/channels.coffee [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+module.exports.init = ->
+  user = new User USER
+
+  new PathView
+    el: $('section.path--header')
+    model: user
+
+  new MetaEditableAttributeView
+    model: user
+    el: $("#metadata--about .metadata__content")
```
Comment (dzucconi) [inline apps/user/client/channels.coffee]: `js-`...

File: apps/user/client/channels.coffee [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+User = require '../../../models/user.coffee'
+PathView = require '../../../components/path/client/path_view.coffee'
+setupChannelsView = require '../components/channels/index.coffee'
+MetaEditableAttributeView = require '../../../components/editable_attribute/client/meta_editable_attribute_view.coffee'
+
+module.exports.init = ->
+  user = new User USER
+
+  new PathView
+    el: $('section.path--header')
```
Comment (dzucconi) [inline apps/user/client/channels.coffee]: `js-`...

File: apps/user/components/channels/collection.coffee (hunk 1 above)
Comment (dzucconi) [inline apps/user/components/channels/collection.coffee]: `@listenTo`

File: apps/user/components/channels/view.coffee [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    @listenTo @collection, 'sync', @render
+    @listenTo @state, 'change', @setLoading
+
+  loadingOrExpired: ->
+    @state.get('loading') or 
+    @state.get('disabled') or 
+    mediator.shared.state.get('lightbox')
+
+  setLoading: ->
+    if @state.get('loading') is true
```
Comment (dzucconi) [inline apps/user/components/channels/view.coffee]: true is true

File: apps/user/components/channels/view.coffee [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+  setUpChannelGroupViews: ->
+    channels = @collection.toJSON()
+    
+    for channel in channels
+      view = new ChannelGroupView
+        channel: channel
+        el: @$(".ChannelBlockGroup[data-id=#{channel.id}]")
+      
+      view.initBlockViews()
```
Comment (dzucconi) [inline apps/user/components/channels/view.coffee]: It's not totally obvious to me but does this view ever get removed? If so there should be a teardown that cleans up the subviews.

File: apps/user/components/channels/view.coffee [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    
+    @setUpChannelGroupViews()
+
+  setUpChannelGroupViews: ->
+    channels = @collection.toJSON()
+    
+    for channel in channels
+      view = new ChannelGroupView
+        channel: channel
+        el: @$(".ChannelBlockGroup[data-id=#{channel.id}]")
```
Comment (dzucconi) [inline apps/user/components/channels/view.coffee]: `js-`

Comment (dzucconi) [inline apps/user/components/channels/view.coffee]: Seriously: you have to do this for me

File: apps/user/components/channels/collection.coffee (hunk 1 above)
Comment (broskoski) [inline apps/user/components/channels/collection.coffee]: AFAIK, there is no `@listenTo` on models or collections

File: apps/user/components/channels/view.coffee (hunk 7 above)
Comment (broskoski) [inline apps/user/components/channels/view.coffee]: Yeah, good call

File: apps/user/components/channels/collection.coffee (hunk 1 above)
Comment (dzucconi) [inline apps/user/components/channels/collection.coffee]: ah woops; you right

File: apps/user/components/channels/view.coffee (hunk 8 above)
Comment (broskoski) [inline apps/user/components/channels/view.coffee]: Okkkkk

**Pull request merged**